### ISSUE TITLE
ServiceBus Session support

### ISSUE THREAD

**Issue by Zenuka** [Azure/azure-webjobs-sdk#529]
I've posted this question yesterday on [StackOverflow](http://stackoverflow.com/questions/30773632/azure-webjob-servicebustrigger-for-sessions) but I guess this is a better place to ask; is it possible to have something like ServiceBusSessionTrigger for handling session through the webjobs sdk?

**Comment by mathewc** [comment]
Can you detail more on the scenarios you're thinking of for it, and how you'd like it to work? The new extensibility model we're opening up would allow you to write one yourself if you wanted to. However, if this is something that many people would find useful, it would make sense for us to bake into our core ServiceBus assembly.

It's not clear to me whether this would need to be a completely new trigger binding, or whether on the trigger/receive side of things it couldn't just be an additional mode on the existing ServiceBusTrigger binding. Behind the scenes, it polls for and accepts sessions, and invokes your job function for session messages.

As far as sending session based messages, I think our ServiceBus binding (e.g. used when sending a BrokeredMessage via out param binding) would work as is? You just set your session ID on the message before sending.

**Comment by Zenuka** [comment]
I'm going to use it for handling the messages of one session in a specific sequence and preferably multiple sessions at once. Since it's currently not possible, I'm going to use [this article](http://www.jayway.com/2014/08/28/routing-commands-using-azure-service-bus-sessions/) to process multiple sessions per webjob.
How should it work? Good question... I think it would be nice to have a method something like this:
```C#
public static void ProcessQueueMessage([ServiceBusTrigger(queueName:"inputqueue", useSessions: true)] List<string> messages) { ... }
```
Or whatever type of object you expect. Also you would need to specify somewhere in the config the timeout between sessions; after how many (milli)seconds will you assume the session is finished?

And you are correct, for sending messages you just need to set something in the SessionId.

**Comment by JasonBSteele** [comment]
I'd love to see this feature as well. I need guaranteed my inbound messages to be in a guaranteed delivery order, and it seems using the Service Bus Session is the way to do this. However it is not at all clear how I would implement this in a WebJob.

**Comment by isaacabraham** [comment]
@mathewc 

Not having this is a blocker for me currently. We're using queues + the webjobs SDK as a very lightweight, scalable reactive work item list - messages come on, and processing those messages leads to 0 .. M new messages going back onto the queue. When we need to have guaranteed sequential processing - almost simulating an actor-style model - service bus sessions is the only option. This is crucial if we have some critical region of data processing and don't want e.g. transactions on SQL database (or can't e.g. Azure blobs).

Having an easy way to target messages on a specific session and have them work transparently within webjobs would be a *huge* benefit and would open a ton of possibilities. So on the input side I would love to see it "just work": -

* Web job trigger has an extra argument to indicate if you want to bind to sessions or just plain messages
* A session is bound to a specific web job process + thread instance and stays with it until the session dries up
* Optionally an extra input argument for the name of the session that has been bound to the function call
* On the output, it would be great if you could specify the session name on the ServiceBus attribute somehow, but this isn't as much of an issue as there's already a way to do this via BrokeredMessage directly as you say.

**Comment by MikeYeager** [comment]
I need this feature as well. We have a WebAPI call that's taking longer and longer as we add more work to each call. When it started to timeout in Azure, we decided to move the process into a WebJob so that clients can initiate the task, then check back to get progress. The clients in this case are browsers making AJAX calls. If we can't correlate the requests with the responses, they can't check on progress or completion of the WebJob. We started looking at Azure Storage Queues, but found that there's no way for the original caller to only receive responses intended for them. We can't create a response queue for each client, because each client is a browser. We saw that Service Bus Queues and Topics allow us to use the SessionId to correlate requests and responses. The client (browser) can filter the response queue to only see responses intended for them. It sounded like a perfect solution. After wrestling with configuration we got the JobHost.RunAndBlock() to almost run, but we get the following error:

**It is not possible for an entity that requires sessions to create a non-sessionful message receiver.**

We found this documentation at http://www.cloudcasts.net/devguide/Default.aspx?id=13029 

If sessions are required on a queue it is not possible to use the QueueClient Receive method to dequeue messages. Attempting to do so will result in.

InvalidOperationException
 
It is not possible for an entity that requires sessions to create a non-sessionful message receiver.

We assume the WebJobs SDK is using the QueueClient Receive method. We're going to try to work around this by having the request queue send a Guid as part of the payload and not require Sessions on the request queue. We'll require sessions on the response queue and see if we can use the Guid in the payload as the SessionId for the response. Will let you know if that works.

**Comment by isaacabraham** [comment]
One thing I've noticed since my OP is that you can use [<Singleton>] on e.g. Storage Queue and use the SingletonScope to achieve virtually the same thing (as far as I can see) as Service Bus sessions. Can anyone confirm that?

**Comment by MikeYeager** [comment]
FYI, the work-around mentioned above worked perfectly:
"having the request queue send a Guid as part of the payload and not requiring Sessions on the request queue. We'll require sessions on the response queue and use the Guid in the payload as the SessionId for the response."

**Comment by Hihaj** [comment]
I would really like to see this implemented as well. I don't use the WebJobs SDK directly that much, but if I understand it correctly this would allow for Azure Functions written in for example JavaScript to handle message sessions (something that's not supported in the Node.js Azure SDK since it uses the REST interface of Azure Service Bus).

**Comment by Bigshooter** [comment]
Was there any traction on this?

**Comment by benjamineberle** [comment]
+1
Our requirement is FIFO by tenant in a multi-tenant system. In this case data imports via upsert

**Comment by brettsam** [comment]
@amit-kumar2 -- is this the issue that you brought up to me yesterday? If so, would you want to weigh in with your scenario here so we have it captured?

**Comment by appalaraju** [comment]
HI,

i am using below code with Servicebus and Topic-Subscription with session enabled

public static void ProcessTopicMessage([ServiceBusTrigger("%ServicebusTopic%", "%ServicebusSubscription%")] BrokeredMessage message, TextWriter logger)
        {
}

but the problem i face with continuously running web job is  host.RunAndBlock(); throws below error so how to resolve this issue? Azure web job does not support Session enabled topic-subscription?


An unhandled exception of type 'System.InvalidOperationException' occurred in mscorlib.dll

Additional information: It is not possible for an entity that requires sessions to create a non-sessionful message receiver. TrackingId:074fbc74-b5e9-4fc2-886b-ed2f4e324304_G5_B15, SystemTracker:az-g3ms-servicebus:Topic:az-rsd-dev-topic|az-rsd-dev-subscription, Timestamp:5/11/2018 5:16:14 AM

**Comment by raulbojalil** [comment]
Hi @appalaraju were you able to resolve your issue?. In my case I'm not getting any exceptions, but the web job is not being fired either. If I disable sessions in my queue the web job starts working again.

**Comment by MikeYeager** [comment]
As far as we have been able to determine, we simply have to accept that the queue will throw exceptions on occasion with no discern-able reason so we just have to handle and ignore it. As far as not working when sessions are enabled, we have not encountered that. Have you tried turning on diagnostics in the portal?

**Comment by mathewc** [comment]
Related to background errors coming from MessageReceivers - we just made some improvements in that area: https://github.com/Azure/azure-webjobs-sdk/issues/1703

**Comment by appalaraju** [comment]
Hi @raulbojalil,MikeYeager

Hhave you tried with Topic + Session enabled Subscription?

in my case, i have created a web job and  used the code(ProcessTopicMessage) which i specified above with Topic and session enabled subscription and ran it locally( in my machine).

i am facing the issue which i specified above.

any idea on when can Azure web job works properly with Topic + Session enabled Subscription?

**Comment by davidrevoledo** [comment]
It'd be good to have this feature :(

**Comment by JasonBSteele** [comment]
So it's been over 2 years since this issue was raised and still nothing on whether/when this will be addressed.

Please vote for it at https://feedback.azure.com/forums/355860-azure-functions/suggestions/13882992-support-for-session-enabled-azure-service-bus-queu

**Comment by davidrevoledo** [comment]
This is a limitation for us, any eta to implement this ?

**Comment by davidrevoledo** [comment]
@ridhoq If you want I join the community support for this

**Comment by JasonBSteele** [comment]
On October 18th a comment was added to https://feedback.azure.com/forums/355860-azure-functions/suggestions/13882992-support-for-session-enabled-azure-service-bus-queu to say the work has been planned... so there is hope!

**Comment by fabiocav** [comment]
Assigning this to sprint 45 for initial design and scoping

**Comment by brettsam** [comment]
Design is ready for internal review. Will be reviewing and planning in Sprint 47. When finalized, we'll respond back here.

**Comment by fabiocav** [comment]
Design work completed. Moving this to sprint 48 to track the implementation

**Comment by alrod** [comment]
Pre-release nuget package is out. Just change the Microsoft.Azure.WebJobs.Extensions.ServiceBus nuget on your VS project if you want to try:
https://myget.org/feed/azure-appservice/package/nuget/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.0.5-11562

Using:
` public static void Run([ServiceBusTrigger("core-test-queue1-sessions", Connection = "AzureWebJobsServiceBus", IsSessionsEnabled = true)]string myQueueItem, ClientEntity clientEntity, ILogger log)`

IsSessionsEnabled = true - if you are pointed to a sessions enabled queue/subscription.
ClientEntity clientEntity - for management operation (like MessageReceiver for a regular queus/subscription). 

Also you can specify new setting sections in host.json:
```
{
    "version": "2.0",
    "extensions": {
        "serviceBus": {
            "SessionHandlerOptions":
             {
                "MaxAutoRenewDuration": "00:01:00",
                "MessageWaitTimeout": "00:05:00",
                "MaxConcurrentSessions": 16,
                "AutoComplete": true,
             }
        }
    }
}
```


I'll appreciate for any feedback.

**Comment by fabiocav** [comment]
Moving this to sprint 49 to allow more time for feedback before this work is merged.

**Comment by appalaraju** [comment]
Is this feature available in Azure functions?

**Comment by appalaraju** [comment]
does this feature supported by topic-subscription?

**Comment by erik-neumann** [comment]
@alrod I cant access the package using your link https://myget.org/feed/azure-appservice/package/nuget/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.0.5-11562 anymore. Few days ago it worked. Any idea?
I was using the ServiceBus extension using the Session support, and it worked pretty well so far.

**Comment by lukasvosyka** [comment]
Having the same issue as @erik-neumann ! Cannot see the Microsoft.Azure.WebJobs.Extensions.ServiceBus nightly builds at all anymore. They seem to be not listed anymore at https://myget.org/gallery/azure-appservice

**Comment by alrod** [comment]
@erik-neumann, @lukasvosyka  please use follwoing link to download the package:
https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta1

**Comment by alrod** [comment]
@agruning, SB sessions feature will be available in PROD later (in May). You can try it using instructions above and it works with topic/subscriptions.

**Comment by ppusateri** [comment]
@alrod - Trying to use 3.1.0 beta 1, but package restoration fails because it depends on Microsoft.Azure.WebJobs >= 3.0.8.  The latest I can see is 3.0.6.  Do we need to wait until 3.0.8 is available to try this?

**Comment by erik-neumann** [comment]
@ppusateri I made it run by additionally referencing Microsoft.Azure.WebJobs 3.0.8 from the myGet feed:
https://myget.org/feed/azure-appservice/package/nuget/Microsoft.Azure.WebJobs/3.0.8

**Comment by paulbatum** [comment]
@ppusateri @erik-neumann sorry about this, I can confirm there's a problem regarding the availability of the 3.0.8 package on NuGet. We're double checking now,

**Comment by fabiocav** [comment]
Again, apologies about this. The packages are up on NuGet and the dependencies should be valid now.

Please let us know if you continue to run into issues.

**Comment by blake-dan** [comment]
Hi, @fabiocav, @alrod 

I have a ServiceBusTrigger function running great where it is getting the message as a string with IsSessionsEnabled = true. However, when I try to bind to the message as a BrokeredMessage I'm getting a deserialization error. The messages do have a session id in the header. Here is the error:

System.Private.CoreLib: Exception while executing function: QueueTrigger. Microsoft.Azure.WebJobs.Host: Exception binding parameter 'message'. System.Private.DataContractSerialization: There was an error deserializing the object of type Microsoft.ServiceBus.Messaging.BrokeredMessage. The input source is not correctly formatted. System.Private.DataContractSerialization: The input source is not correctly formatted.

Here is my function definition:
        [FunctionName("QueueTrigger")]
        public static void Run(
            [ServiceBusTrigger("%AsbQueueName%", Connection = "AsbQueueConnection", IsSessionsEnabled = true)] BrokeredMessage message,
            ILogger log)
        {

Should this work?
Thanks!

**Comment by JasonBSteele** [comment]
> Here is my function definition:
> [FunctionName("QueueTrigger")]
> public static void Run(
> [ServiceBusTrigger("%AsbQueueName%", Connection = "AsbQueueConnection", IsSessionsEnabled = true)] BrokeredMessage message,
> ILogger log)

The example above uses a string:
`public static void Run([ServiceBusTrigger("core-test-queue1-sessions", Connection = "AzureWebJobsServiceBus", IsSessionsEnabled = true)]string myQueueItem, ClientEntity clientEntity, ILogger log)`

Also `BrokeredMessage` is from the old Service Bus nuget package which this may not be compatible with.

**Comment by appalaraju** [comment]
Hi ,


I just want your confirmation whether "ServiceBus Session" support my requirement or not?

my requirement is 
there are multiple devices which processes cloud requests concurrently and provide responses to the each cloud request with some responses in the specific order like ( In-progress,pending,going-on,success) at the same time.
Please look into below example

1. Let us say there are 2 devices ( ex: DV1 and DV2).
2. Each device sends response messages to IOT Hub in sequential/ordered manner at the same time.
 ( ex :  DV1-In-progress,DV1-pending,DV1-going-on,DV1-success).
 ( ex : DV2-In-progress,DV2-pending,DV2-going-on,DV2-success).
3. Each brokeredmessage will be sent to cloud with unique session key.
4. Let us say there are 2 web jobs running in the cloud with  IsSessionsEnabled = true
5. So if messages from different devices(ex: DV1 and DV2) reach topic-subscription at the same time there is chance that topic-subscription will be filled in the order of jumbling like ( ex :  DV1-In-progress,DV2-In-progress,DV1-pending,DV1-going-on,DV2-pending,DV1-success,DV2-going-on,DV2-success)
6. So here 2 web jobs try read the messages from topic at the same time
7. So as per my understanding, "IsSessionsEnabled = true" concept enables one web job to read Device specific messages in the same order sent by one device(ex: DV1)  ( ex: DV1-In-progress,DV1-pending,DV1-going-on,DV1-success) and other web job reads DV2 messages ( ex: DV2-In-progress,DV2-pending,DV2-going-on,DV2-success)
8. is my understanding correct?

Thanks,
Raju

**Comment by JasonBSteele** [comment]
> 3\. Each brokeredmessage will be sent to cloud with unique session key.

You need to set the SessionId of the message to the device Id so that the messages you want to keep in sequence with each other share the same SessionId.

**Comment by alrod** [comment]
@blake-dan, SB Sessions is a V2 runtime feature. You should use Message instead of BrokeredMessage:
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus#trigger---usage

**Comment by alrod** [comment]
@appalaraju, 
@JasonBSteele is right.

If you want guaranteed ServiceBusTrigger processing sequence "DV1-In-progress,DV1-pending,DV1-going-on,DV1-success", DV1 must send the messages with the same SessinoId and with the order "DV1-In-progress,DV1-pending,DV1-going-on,DV1-success"

**Comment by appalaraju** [comment]
@JasonBSteele , @alrod --> thanks for your confirmation and valuable details.

**Comment by appalaraju** [comment]
Hi All,

I have few more questions as below.Please look into them.

1. I am using topic-subscription so web jobs read messages from the defined topic-subscription. there is option called "Enable Partitioning" for a topic so to make "ServiceBus Session" concept works properly, should "Enable Partitioning" option be set to "True" or "False"?   
2. Is there any document published by Microsoft team to understand "ServiceBus Session" concept more clearly?

please look into below example.

1) Let us say there are 2 devices ( ex: DV1 and DV2).
2) Each device sends response messages to IOT Hub in sequential/ordered manner at the same time.
( ex : DV1-In-progress,DV1-pending,DV1-going-on,DV1-success).
( ex : DV2-In-progress,DV2-pending,DV2-going-on,DV2-success).
3) all brokeredmessage coming from same device will be sent to cloud with same session key in order to maintain the sequence.
4) Let us say there is 1 web job running in the cloud with IsSessionsEnabled = true
5) So if messages from different devices(ex: DV1 and DV2) reach topic-subscription at the same time then there is chance that topic-subscription will be filled with messages in the order of jumbling like ( ex : DV1-In-progress,DV2-In-progress,DV1-pending,DV1-going-on,DV2-pending,DV1-success,DV2-going-on,DV2-success)
6) So here 1 web job try read the messages from topic.
7) So as per my understanding, "IsSessionsEnabled = true" concept enables that web job to read Device specific messages in the same order sent by one device(ex: DV1) ( ex: DV1-In-progress,DV1-pending,DV1-going-on,DV1-success) so after reading DV1 messages then only that web job reads DV2 messages ( ex: DV2-In-progress,DV2-pending,DV2-going-on,DV2-success)
8) is my understanding correct?

**Comment by DAllanCarr** [comment]
Yes. Partitioned queues use SessionId as PartitionKey if not set. Partition queues are not required. But the high consistency is reduced as messaging operations are always routed to a specific partition.

**Comment by JasonBSteele** [comment]
> 7\. So as per my understanding, "IsSessionsEnabled = true" concept enables that the single web job firstly reads all same Device specific messages in the same order sent by one device(ex: DV1) ( ex: DV1-In-progress,DV1-pending,DV1-going-on,DV1-success) so after reading DV1 messages then only that web job reads DV2 messages ( ex: DV2-In-progress,DV2-pending,DV2-going-on,DV2-success)

This is my understanding...

By default your WebJob or Function can have multiple instances running. An instance will be created to handle the first DV1 message, then another will be created to handle the first DV2 messages. While these instances hold on to the Session Lock they will continue to receive the messages for the relevant devices. 

If they don't receive a message from a device for a while they will release the Session Lock and be shut down. When a new message for the device eventually does arrive a new instance is created that will continue to receive messages for that device until it too is shut down.

**Comment by brotaru89** [comment]
Using the trigger with MessageReceiver does not work when IsSessionsEnabled is set to True.
Example: 
```
([ServiceBusTrigger("my.queue.with.sessions", Connection = "ServiceBusEndpoint", IsSessionsEnabled = true)]string myQueueItem, string messageId, ILogger log, MessageReceiver messageReceiver, string lockToken)
```

The "messageReceiver" variable is NULL. Can you please have a look? Thanks!

**Comment by DAllanCarr** [comment]
@brotaru89  I think this is intentional you will need to bind to `ClientEntity` and cast.

**Comment by alrod** [comment]
@appalaraju


> I have few more questions as below.Please look into them.
> 
> 1. I am using topic-subscription so web jobs read messages from the defined topic-subscription. there is option called "Enable Partitioning" for a topic so to make "ServiceBus Session" concept works properly, should "Enable Partitioning" option be set to "True" or "False"?
> 2. Is there any document published by Microsoft team to understand "ServiceBus Session" concept more clearly?

Enable Partitioning is a separate feature and does not affect SB sessions:
https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-partitioning

Document for SB Sessions:
https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions

**Comment by brotaru89** [comment]
@blake-dan Let's assume that the function app calls an external API (via HTTP) and based on the API response it needs to take the decision if the message should be completed or sent to dead-letter based on some business rules, but without throwing an explicit exception.

Moreover, setting the configuration flag **_autoComplete = false_** should enable you to manually decide how to manage/handle the message.

**Comment by brotaru89** [comment]
@blake-dan Let's assume that the function app needs to call an external API (via HTTP call) and based on the API response it should decide whether to complete the message or send it to dead-letter queue. This should be possible without throwing exceptions explicitly and thus, failing the function execution.

Moreover, setting the runtime configuration flag **"autoComplete": false** should allow you to manually manage/handle the messages.

**Comment by brotaru89** [comment]
@blake-dan Let's assume that the function app needs to call an external API (via HTTP call) and based on the API response it should decide whether to complete the message or send it to dead-letter queue. This should be possible without throwing exceptions explicitly and thus, failing the function execution.

Moreover, setting the runtime configuration flag **"autoComplete": false** should allow you to manually manage/handle the message

**Comment by brotaru89** [comment]
@blake-dan Let's assume that the function app needs to call an external API (via HTTP call) and based on the API response it should decide whether to complete the message or send it to dead-letter queue. This should be possible without throwing exceptions explicitly and thus, failing the function execution.

Moreover, setting the runtime configuration flag **"autoComplete": false** should allow you to manually manage/handle the message.

**Comment by brotaru89** [comment]
.

**Comment by brotaru89** [comment]
> > @alrod Thanks for your answer regarding PR 2199
> > Hence, I've changed my function app signature to include a ClientEntity parameter, which is then cast to QueueClient. Further on, I'm using the queue client to mark the message as completed or dead-lettered depending on some business rules. But, the QueueClient.CompleteAsync or QueueClient.DeadLetterAsync are failing with the following error message:
> > System.Private.CoreLib: Exception while executing function: FileWatcherFunction. Microsoft.Azure.ServiceBus: It is not possible for an entity that requires sessions to create a non-sessionful message receiver. TrackingId:7f443619-258e-4d45-848f-4f1bb62c473d_B1, SystemTracker:my-mq-testing:queue:my.queue.with.sessions~15, Timestamp:2019-05-08T12:00:44 TrackingId:de193676dbf44b24a9f6ce9bd58b530c_G57, SystemTracker:gateway7, Timestamp:2019-05-08T12:00:44.
> > Please note that the queue has sessions enabled. Please see code sample below:
> > [FunctionName("Function1")]
> > public static async Task Run(
> > [ServiceBusTrigger("my.queue.with.sessions", Connection = "ServiceBusEndpoint", IsSessionsEnabled = true)]
> > string myQueueItem, string messageId, ILogger log, ClientEntity clientEntity, string lockToken)
> > {
> > log.LogInformation($"[{messageId}] > Function started");
> > log.LogInformation($"[{messageId}] > Message content is [{myQueueItem}]");
> > ```
> > var queueClient = clientEntity as QueueClient;
> > if (queueClient is null)
> >     log.LogError($"Type [{clientEntity.GetType()}] is not handled/supported.");
> > 
> > await queueClient.CompleteAsync(lockToken);
> > ```
> > 
> > 
> > }
> > @alrod @fabiocav Can you please advise?
> 
> @brotaru89,
> Complete (on the message) is performed automatically if the function run is successful. Abandon is performed if an exception occurs in the function. Dead letter happens after a (configurable) number of abandoned locks. As I understand it, the MessageWaitTimeout app setting controls how long Azure Service Bus waits for new messages to enter a session before it closes the session. This is all handled automatically. No need to handle complete, dead letter in your function.

@blake-dan Let's assume that the function app needs to call an external API (via HTTP call) and based on the API response it should decide whether to complete the message or send it to dead-letter queue. This should be possible without throwing exceptions explicitly and thus, failing the function execution.

Moreover, setting the runtime configuration flag **"autoComplete": false** should allow you to manually manage/handle the message

**Comment by blake-dan** [comment]
> @alrod Thanks for your answer regarding PR 2199
> Hence, I've changed my function app signature to include a ClientEntity parameter, which is then cast to QueueClient. Further on, I'm using the queue client to mark the message as completed or dead-lettered depending on some business rules. But, the QueueClient.CompleteAsync or QueueClient.DeadLetterAsync are failing with the following error message:
> 
> System.Private.CoreLib: Exception while executing function: FileWatcherFunction. Microsoft.Azure.ServiceBus: It is not possible for an entity that requires sessions to create a non-sessionful message receiver. TrackingId:7f443619-258e-4d45-848f-4f1bb62c473d_B1, SystemTracker:my-mq-testing:queue:my.queue.with.sessions~15, Timestamp:2019-05-08T12:00:44 TrackingId:de193676dbf44b24a9f6ce9bd58b530c_G57, SystemTracker:gateway7, Timestamp:2019-05-08T12:00:44.
> 
> Please note that the queue has sessions enabled. Please see code sample below:
> [FunctionName("Function1")]
> public static async Task Run(
>     [ServiceBusTrigger("my.queue.with.sessions", Connection = "ServiceBusEndpoint", IsSessionsEnabled = true)]
>     string myQueueItem, string messageId, ILogger log, ClientEntity clientEntity, string lockToken)
> {
>     log.LogInformation($"[{messageId}] > Function started");
>     log.LogInformation($"[{messageId}] > Message content is [{myQueueItem}]");
> 
>     var queueClient = clientEntity as QueueClient;
>     if (queueClient is null)
>         log.LogError($"Type [{clientEntity.GetType()}] is not handled/supported.");
> 
>     await queueClient.CompleteAsync(lockToken);
> }
> 
> @alrod @fabiocav Can you please advise?

@brotaru89,
Complete (on the message) is performed automatically if the function run is successful. Abandon is performed if an exception occurs in the function. Dead letter happens after a (configurable) number of abandoned locks. As I understand it, the MessageWaitTimeout app setting controls how long Azure Service Bus waits for new messages to enter a session before it closes the session. This is all handled automatically. No need to handle complete, dead letter in your function.

**Comment by DAllanCarr** [comment]
`CompleteAsync()` method can only be used for a non-sessionful subscription.

**Comment by alrod** [comment]
@brotaru89, it looks like `IMessageSession` should be exposed instead of `ClientEntity` for sessions management operations. 
3.1.0-beta2 nuget package with `IMessageSession` support will be uploaded soon.

**Comment by alrod** [comment]
https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 is out. 
Difference with 3.1.0-beta1 is using `IMessageSession messageSession` instead of `ClientEntity clientEntity`.

Using:
` public static void Run([ServiceBusTrigger("core-test-queue1-sessions", Connection = "AzureWebJobsServiceBus", IsSessionsEnabled = true)]string myQueueItem, IMessageSession messageSession, ILogger log)`

`IsSessionsEnabled = true` - if you are pointed to a sessions enabled queue/subscription.
`IMessageSession messageSession` - for management operations (like MessageReceiver for a regular queue/subscription). 

Also you can specify new `SessionHandlerOptions` section in host.json:
```
{
    "version": "2.0",
    "extensions": {
        "serviceBus": {
            "SessionHandlerOptions":
             {
                "MaxAutoRenewDuration": "00:01:00",
                "MessageWaitTimeout": "00:05:00",
                "MaxConcurrentSessions": 16,
                "AutoComplete": true,
             }
        }
    }
}
```

**Comment by brotaru89** [comment]
@alrod Thanks for your support and prompt response. I've tested the beta2 version and it works as expected.

**Comment by appalaraju** [comment]
Hi,

I am using below code in my web job to read messages from topic and subscription.
I have created a subscription with "Sessions Enabled=true" so 
Is this enough for web job to read session enabled messages from topic and subscription?
 or 
Need to set "IsSessionsEnabled=true" for ProcessTopicMessage function also in web job? if yes, how to set "IsSessionsEnabled=true" for ProcessTopicMessage function?

below is the code ProcessTopicMessage function.


 public async static Task ProcessTopicMessage([ServiceBusTrigger("%ServicebusRSDTopic%", "%ServicebusRSDSubscription%")] BrokeredMessage message, TextWriter logger,
        CancellationToken cancellationToken)
{

}

**Comment by alrod** [comment]
@appalaraju, I see `BrokeredMessage` in your code. Is your function app V1 or V2? The sessions feature works only for V2 function apps + you need beta package:
https://github.com/Azure/azure-webjobs-sdk/issues/529#issuecomment-491113458

In your case the code should be something like this:
`public static void ProcessTopicMessage([ServiceBusTrigger(""%ServicebusRSDTopic%", "%ServicebusRSDSubscription%", IsSessionsEnabled = true)]Message myItem, ILogger log)`

**Comment by jeffhollan** [comment]
Moving to epic and we'll sort out where to track the work to move this from `beta` to production, but for those interested in using the feature today it is on NuGet now https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2

**Comment by appalaraju** [comment]
@alrod, @ jeffhollan

I have below points so please look

 into them

1. As you suggested, i have installed https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 in my azure web job.
2. After installing and after compiling my web job, i see errors shown in program_errors.png
![Program_errors](https://user-images.githubusercontent.com/17615932/57827583-7f543b00-77c5-11e9-9894-640014f6d5aa.png)
3. As suggested, i have added "IsSessionsEnabled = true" to ProcessTopicMessage function available in Functions.cs but it is throwing error as ProcessTopicMessage is not accepting "IsSessionsEnabled = true"
![Functions_errors](https://user-images.githubusercontent.com/17615932/57827640-ac085280-77c5-11e9-93d0-8a376deaa087.png)

**Comment by marxxxx** [comment]
Will this also solve this issue? https://github.com/Azure/azure-functions-host/issues/2721

**Comment by appalaraju** [comment]
> Moving to epic and we'll sort out where to track the work to move this from `beta` to production, but for those interested in using the feature today it is on NuGet now https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2

@ jeffhollan , can you please look into my queries specified above. 3.1.0-beta2 is throwing issues as specified above

**Comment by appalaraju** [comment]
> Will this also solve this issue? [Azure/azure-functions-host#2721](https://github.com/Azure/azure-functions-host/issues/2721)

@marxxxx you are talking about the issues i specified above?

**Comment by marxxxx** [comment]
@appalaraju i was referring to the situation that the Azure Service Bus Function trigger currently constantly polls Azure Service Bus in order to find out if there are messages to process. This causes lots of requests to the service and in turn increased cost even if there are no messages to process. I was wondering if session support is related to this situation and could possible solve the problem.

**Comment by appalaraju** [comment]
> https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 is out.
> Difference with 3.1.0-beta1 is using `IMessageSession messageSession` instead of `ClientEntity clientEntity`.
> 
> Using:
> ` public static void Run([ServiceBusTrigger("core-test-queue1-sessions", Connection = "AzureWebJobsServiceBus", IsSessionsEnabled = true)]string myQueueItem, IMessageSession messageSession, ILogger log)`
> 
> `IsSessionsEnabled = true` - if you are pointed to a sessions enabled queue/subscription.
> `IMessageSession messageSession` - for management operations (like MessageReceiver for a regular queue/subscription).
> 
> Also you can specify new `SessionHandlerOptions` section in host.json:
> 
> ```
> {
>     "version": "2.0",
>     "extensions": {
>         "serviceBus": {
>             "SessionHandlerOptions":
>              {
>                 "MaxAutoRenewDuration": "00:01:00",
>                 "MessageWaitTimeout": "00:05:00",
>                 "MaxConcurrentSessions": 16,
>                 "AutoComplete": true,
>              }
>         }
>     }
> }
> ```

@alrod --> after installing Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 into Azure web job sample project, Run function is not accepting `IsSessionsEnabled = true` , am i missing any thing?

**Comment by M05pr1mty** [comment]
Great feature - thanks.

How can I filter my function for a Queue to only pick up specific SessionId values. I would like to deploy a function-per-sessionId.

Or am I overthinking this? And function executions will occur in parallel for sessions.

**Comment by DAllanCarr** [comment]
You would only need a single function to obtain the session lock and configure `MaxConcurrentSessions`

**Comment by alrod** [comment]
> @appalaraju i was referring to the situation that the Azure Service Bus Function trigger currently constantly polls Azure Service Bus in order to find out if there are messages to process. This causes lots of requests to the service and in turn increased cost even if there are no messages to process. I was wondering if session support is related to this situation and could possible solve the problem.

@marxxxx, the feature does not fix "constantly polls Azure Service Bus in order to find out if there are messages to process" issue.

**Comment by alrod** [comment]
> Great feature - thanks.
> 
> How can I filter my function for a Queue to only pick up specific SessionId values. I would like to deploy a function-per-sessionId.
> 
> Or am I overthinking this? And function executions will occur in parallel for sessions.

You do not need to deploy "function-per-sessionId".
The SB binding does not support "sessionId" as parameter. So all sessions will trigger one function app.
You can do specific actions depending on 'sessionId' (you can get sessionId as Message property) inside your function app code.

**Comment by alrod** [comment]
> @alrod --> after installing Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 into Azure web job sample project, Run function is not accepting `IsSessionsEnabled = true` , am i missing any thing?

Can you please upload a zip of your project somewere? Then I can take a look.

**Comment by appalaraju** [comment]
> > @alrod --> after installing Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2 into Azure web job sample project, Run function is not accepting `IsSessionsEnabled = true` , am i missing any thing?
> 
> @appalaraju, Can you please upload a zip of your project somewere? Then I can take a look.

@alrod , my actual project was referring/using "Microsoft.Azure.WebJobs.ServiceBus( V 2.0.0) so i uninstalled it and installed Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2. now there is no issues with "IsSessionsEnabled = true".

so my question here is 
we should not use "Microsoft.Azure.WebJobs.ServiceBus( V 2.0.0)" if we want to use "Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2" ?

and 
i am facing 1 more issue which is getting below issue after installing "Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2"
 "Error	CS0246	The type or namespace name 'JobHostConfiguration' could not be found (are you missing a using directive or an assembly reference?) "

so can not we use JobHostConfiguration if we want to use "Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta2" ?

**Comment by alrod** [comment]
@appalaraju, what version of webjobs-sdk-versions do you use?
https://docs.microsoft.com/en-us/azure/app-service/webjobs-sdk-how-to#webjobs-sdk-versions
you should be referenced to 3.x

**Comment by appalaraju** [comment]
> @appalaraju, what version of webjobs-sdk-versions do you use?
> https://docs.microsoft.com/en-us/azure/app-service/webjobs-sdk-how-to#webjobs-sdk-versions
> you should be referenced to 3.x

@alrod , i am using V3.0.8 of Microsoft.Azure.WebJobs as specified below
![image](https://user-images.githubusercontent.com/17615932/58148292-c639a900-7c7b-11e9-8db3-cedbb8e060d8.png)

**Comment by alrod** [comment]
@appalaraju your project is targeting .NET Framework 4.6.1. You need to create a .NET Core 2.2 project.

**Comment by appalaraju** [comment]
> @appalaraju your project is targeting .NET Framework 4.6.1. You need to create a .NET Core 2.2 project.

@alrod , Thanks for your valuable information. i will try and reach you.

**Comment by ahocquet** [comment]
I've tried the 3.1.0-beta2 nuget package but I'm having the following error : 
```
A host error has occurred
Microsoft.Azure.WebJobs.ServiceBus: Index was outside the bounds of the array.
```

I have the following configuration:

- Visual Studio 2019 V16.1.1
- .NET Core 2.2 Project
- Microsoft.NET.Sdk.Functions v1.0.27

I've uploaded my project to this repo : https://github.com/ahocquet/azure-function-service-bus-session

Some screenshots:
![](https://i.imgur.com/NXgEYYz.png)
![](https://i.imgur.com/1bIuOF6.png)

**Comment by emi662002** [comment]
I'm using functions v2, with IsSessionsEnabled=true
I have a Topic+Subscription with session enabled
But my function is not triggering.
It does trigger when I use a topic+subsription with session disabled and IsSessionDIsabled=false. 

VS2019
Microsoft.Azure.WebJobs.Extensions.ServiceBus 3.1.0-beta2
.net core 2.2
sdk functions 1.0.28

**Comment by jeffhollan** [comment]
Do you have a session enabled topic? Only one mode will work per queue, so if the same topic works when sessions is turned off the topic likely wasn’t created to use sessions.

—
Jeff
________________________________

**Comment by emi662002** [comment]
It is session enabled indeed. I've meant that when I create a (different)
session disabled one the rest of the code does work, so the code, including
secrets,etc it's working, except for sessions.

Op zo 2 jun. 2019 21:26 schreef Jeff Hollan <notifications@github.com>:

**Comment by jeffhollan** [comment]
Hmm not sure on this. Do you see any errors being thrown or does it just not trigger?

**Comment by emi662002** [comment]
> Hmm not sure on this. Do you see any errors being thrown or does it just not trigger?

it just won't trigger. I don't get any errors, the other functions on my solution trigger just ok (those are not session enabled) 
@jeffhollan I've followed your example solution but this one is with queues, and not topic+subscription. would this be the problem? I know with t+s you have queues, but it's not the same trigger configuration....

**Comment by alrod** [comment]
> > Hmm not sure on this. Do you see any errors being thrown or does it just not trigger?
> 
> it just won't trigger. I don't get any errors, the other functions on my solution trigger just ok (those are not session enabled)
> @jeffhollan I've followed your example solution but this one is with queues, and not topic+subscription. would this be the problem? I know with t+s you have queues, but it's not the same trigger configuration....

@emi662002, can you please share your project with the no-working trigger? or can you please post the definition of your trigger?

**Comment by alrod** [comment]
> I've tried the 3.1.0-beta2 nuget package but I'm having the following error :
> 
> ```
> A host error has occurred
> Microsoft.Azure.WebJobs.ServiceBus: Index was outside the bounds of the array.
> ```
> 
> I have the following configuration:
> 
> * Visual Studio 2019 V16.1.1
> * .NET Core 2.2 Project
> * Microsoft.NET.Sdk.Functions v1.0.27
> 
> I've uploaded my project to this repo : https://github.com/ahocquet/azure-function-service-bus-session
> 
> Some screenshots:
> ![](https://camo.githubusercontent.com/eef33704440ef1e30f3ed9f6faba97ca541d7a77/68747470733a2f2f692e696d6775722e636f6d2f4e58674559597a2e706e67)
> ![](https://camo.githubusercontent.com/d8214bf1747e05a27461684ad1e9ae1af5082f90/68747470733a2f2f692e696d6775722e636f6d2f316249754f46362e706e67)

I have identified the bug: session trigger does not work when queue name contains "/". It will be fixed in next beta version.

**Comment by emi662002** [comment]
> > > Hmm not sure on this. Do you see any errors being thrown or does it just not trigger?
> > 
> > 
> > it just won't trigger. I don't get any errors, the other functions on my solution trigger just ok (those are not session enabled)
> > @jeffhollan I've followed your example solution but this one is with queues, and not topic+subscription. would this be the problem? I know with t+s you have queues, but it's not the same trigger configuration....
> 
> @emi662002, can you please share your project with the no-working trigger? or can you please post the definition of your trigger?

 [FunctionName(nameof(OrderedFunction))]
        public static async Task Run([ServiceBusTrigger("orderedtopic",
                                   "orderedbread", Connection = "ReadConnection",IsSessionsEnabled = true)]string mySbMsg,
                                     ILogger log)

let me know if you need something else, and thanks 👍

**Comment by alrod** [comment]
@emi662002, your trigger definition looks good. I tested the same trigger definition and it works for me.
Are you testing locally or in Azure? 
Can you please share your function name if you test in Azure:
https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Comment by emi662002** [comment]
> @emi662002, your trigger definition looks good. I tested the same trigger definition and it works for me.
> Are you testing locally or in Azure?
> Can you please share your function name if you test in Azure:
> https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

2019-06-03T19:45:31.658 [Information] Executing 'OrderedFunction' (Reason='This function was programmatically called via the host APIs.', Id=f2d835cc-f00d-4a13-ba7e-94dcf0a7503e)
West Europe.

Locally and on Azure both are not triggering. On azure when I run according to your link I see a null pointer exception.

**Comment by alrod** [comment]
@emi662002
Azure:
1. Your function is triggered on Azure as you see in the logs:
[Information] Executing 'OrderedFunction' (Reason='This function was programmatically called via the host APIs.', Id=f2d835cc-f00d-4a13-ba7e-94dcf0a7503e)
2. You got an unhandled "System.NullReferenceException" in your function code:
```
   at async AppCuesWorker.CosmosRepo.Insert(Bread pew,ILogger log) at D:\agent\_work\453\s\AppCuesWorker\CosmosRepo.cs : 22
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async AppCuesWorker.OrderedFunction.Run(String mySbMsg,ILogger log,ISample sample) at D:\agent\_work\453\s\AppCuesWorker\OrderedFunction.cs : 28
```

Locally:
It is possible the SB message is triggered by the function on Azure so local function does not trigger if you have both up and running in the same time.
If it's not the case please compare "ReadConnection" setting locally and on Azure.

**Comment by emi662002** [comment]
> @emi662002
> Azure:
> 
> 1. Your function is triggered on Azure as you see in the logs:
>    [Information] Executing 'OrderedFunction' (Reason='This function was programmatically called via the host APIs.', Id=f2d835cc-f00d-4a13-ba7e-94dcf0a7503e)
> 2. You got an unhandled "System.NullReferenceException" in your function code:
> 
> ```
>    at async AppCuesWorker.CosmosRepo.Insert(Bread pew,ILogger log) at D:\agent\_work\453\s\AppCuesWorker\CosmosRepo.cs : 22
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at async AppCuesWorker.OrderedFunction.Run(String mySbMsg,ILogger log,ISample sample) at D:\agent\_work\453\s\AppCuesWorker\OrderedFunction.cs : 28
> ```
> 
> Locally:
> It is possible the SB message is triggered by the function on Azure so local function does not trigger if you have both up and running in the same time.
> If it's not the case please compare "ReadConnection" setting locally and on Azure.

@alrod When i test locally, I disable the function on azure. Strange thing is that my breakpoint is never hit.

The null pointer on Azure seems to come from an empty string message when the function is invoked. Because I deserialize it and then use the object that nulled is.

**Comment by alrod** [comment]
@ahocquet,
https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta3 is out.
"Session trigger does not work if queue name contains '/' " is fixed this package.

**Comment by ahocquet** [comment]
@alrod 

Thanks for the update! Unfortunately, it seems broken at the moment as it tries to download Microsoft.Azure.Webjobs 3.0.9 that is still not available on nuget.

![image](https://user-images.githubusercontent.com/1451308/59082706-bfa26700-88f4-11e9-92be-6f74ba42e533.png)

**Comment by alrod** [comment]
@ahocquet, just pushed 3.0.9 to nuget. thanks!

**Comment by ahocquet** [comment]
That was fast, thank you! I can confirm that now it's working as expected, event when the queue has a "/" in its name. 🎉

**Comment by KyleLehtinenDev** [comment]
Having problems getting a sessionful message reciever setup. Have closely reviewed the discussion here but not sure where my problem is. 

Getting this error message when trying to run a dotnet azure function locally with a ServiceBusTrigger (stripped identifying info):
`[6/25/19 8:37:12 PM] Message processing error (Action=Receive, ClientId=MessageReceiver1XXXX, EntityPath=XXXXX, Endpoint=XXXXX.servicebus.windows.net)
[6/25/19 8:37:12 PM] Microsoft.Azure.ServiceBus: It is not possible for an entity that requires sessions to create a non-sessionful message receiver. TrackingId:d6569327-0b34-48a4-a24c-0554bf7a0de9_B20, SystemTracker:XXXXX:Queue:XXXXXX, Timestamp:2019-06-25T20:37:12 TrackingId:0f6053cf61ef4455a98b90521bb58523_G15, SystemTracker:gateway7, Timestamp:2019-06-25T20:37:12.`

I'm using netcore2.2, 3.1.0-beta3 of the webjobs extension, and have passed in IsSessionsEnabled = true.

My code is mainly just from getting started docs:
```
using System;
using Microsoft.Azure.ServiceBus;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Host;
using Microsoft.Extensions.Logging;

namespace Company.Function
{
    public static class ServiceBusQueueTriggerCSharp
    {
        [FunctionName("ServiceBusQueueTriggerCSharp")]
        public static void Run(
          [ServiceBusTrigger("testsessionqueue", Connection = "corp-us2-apigw_SERVICEBUS", IsSessionsEnabled = true)]
            string myQueueItem, 
          IMessageSession session,
          ILogger log)
        {
            log.LogInformation($"C# ServiceBus queue trigger function processed message: {myQueueItem}");
        }
    }
}

```
"testsessionqueue" is a session queue. Any suggestions?

**Comment by ahocquet** [comment]
When using a consumption plan for my Azure Function, it doesn't seem to scale out. I only get one instance of my Azure Function running, even though I have hundreds of messages in my queue and my CPU is bound to 100% usage.

Do you guys need a sample to reproduce the issue or are you aware of it?

**Comment by tidusjar** [comment]
Every time my message completes using 3.1.0-beta3 and 3.0.9 i'm getting the following:
 > Microsoft.Azure.ServiceBus: The session lock has expired on the MessageSession. Accept a new MessageSession.

**Comment by tidusjar** [comment]
One thing that does seem to be a regression is that with 3.1.0-beta3 I can no longer use the `%Variable%` in the topic name, I just get the following:
> Microsoft.Azure.ServiceBus: Put token failed. status-code: 404, status-description: The messaging entity 'sb://my-service-bus.servicebus.windows.net/%25AzureServiceBusTopicName%25/Subscriptions/mysub' could not be found. TrackingId:20fa196e-57f4-468d-a727-e2f6f997b208_G11, SystemTracker:my-service-bus.servicebus.windows.net:%25AzureServiceBusTopicName%25/Subscriptions/mysub, Timestamp:2019-07-05T07:20:45.

My code:

```
 public static async Task Run([ServiceBusTrigger("%AzureServiceBusTopicName%", "mysub", Connection = "AzureServiceBusPrimary", IsSessionsEnabled = true)]
            Message message, IMessageSession messageSession, ExecutionContext executionContext, ILogger log)

```

**Comment by mpnow** [comment]
> I then encountered an error about being unable to bind the IMessageSession parameter. When I commented that out the code above processed messages in the queue. I then changed my sample code "session" to "messageSession" and it worked

Same here - when I try any other variable name than "messageSession" I get the "unable to bind" error; using "messageSession", it works. This is with 3.1.0-beta3.

**Comment by mpnow** [comment]
For my own use case, I'd like to have a function that processes in parallel for different session IDs - that is, starts and runs a new instance of the function each time it encounters a different session ID in the input service bus queue, so multiple instances might be running and processing simultaneously - but operates in series for messages of the same session ID - that is, if it encounters a message of a session ID that's currently processing, it waits to start processing the queued message until the currently processing item of that session ID is done. Will this new session support work this way by default, or do I need to use the [Singleton] attribute on the function, or take another approach?

**Comment by alrod** [comment]
@tidusjar, @mpnow I am investigating `%Variable%` issue in beta3. I will update the thread.

**Comment by alexgman** [comment]
If we have a servicebus-triggered function, how do we process all the messages for that specific session?

**Comment by alrod** [comment]
@mpnow, @KyleLehtinenDev. Regarding `IMessageSession  messageSession`, the parameter name must have exact name `messageSession` as it is a [trigger metadata value](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-expressions-patterns#trigger-metadata).

**Comment by alrod** [comment]
> If we have a servicebus-triggered function, how do we process all the messages for that specific session?

There is no support for filtering by sessionId using trigger parameters. What you can do:

1. In host.json `SessionHandlerOptions.AutoComplete = false`
2. In your function code: 
```
if (message.SessionId == "value") {
// Processing logic..
messageSession.CompleteAsync(message.SystemProperties.LockToken);
}
```
Note the trigger will be constantly firing if the entity contains a message with `sessionId != "value"`.

**Comment by rybczak** [comment]
When can we expect to have this feature released to production?

**Comment by alrod** [comment]
https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/3.1.0-beta4 is out.

@tidusjar, @mpnow the issue for resolving %Variable% is fixed in beta4.
@rybczak, most likely the feature will be released in September
